python/lsst/ts/dream/csc/dream_csc.py

```diff
@@ -132,43 +178,129 @@ async def begin_disable(self, id_data: salobj.BaseDdsDataType) -> None:
         id_data: `CommandIdData`
             Command ID and data
         """
+        await self.disconnect()
         await self.cmd_disable.ack_in_progress(id_data, timeout=60)
         await super().begin_disable(id_data)
 
     async def disconnect(self) -> None:
         """Disconnect the DREAM CSC, if connected."""
         self.log.info("Disconnecting")
-        await self.model.disconnect()
+        self.weather_sleep_task.cancel()
+        await self.weather_loop_task
+        if self.model is not None:
+            await self.model.disconnect()
+        self.model = None
+        if self.mock:
+            await self.mock.close()
+            self.mock = None
+        self.log.info("Disconnected.")
```

Comment:
Please make sure to disconnect from ESS here too. You only cancel the `weather_loop_task` but keep the remote connected.